### PR TITLE
Add dependency groups to `pyproject.toml`

### PR DESCRIPTION
**Pull request by tobiasdiez**
Following [PEP 735](https://peps.python.org/pep-0735/#canonical-names-of-dependencies-and-dev-el-split-packages), add a few groups of dependencies that one usually needs while developing sage (e.g. for tests or docs).

This info will then be used to generate the conda lock files and once support for PEP 735 is more widespread other tools will be able to use the metadata as well. Eg, once https://github.com/pypa/pip/issues/12963 is implemented one can install all test dependencies with `pip install --groups test`.

### :memo: Checklist

- [ ] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dimpase** [comment]
https://github.com/sagemath/sage/pull/37482 - is a peculiar PR, just cherry-picking commits from here, a PR on which I can't even comment, and positively reviewed by the PR author himself...

**Comment by tobiasdiez** [comment]
> #37482 - is a peculiar PR, just cherry-picking commits from here, a PR on which I can't even comment, and positively reviewed by the PR author himself...

It's not even a cherry-picked commit. Matthias copied a subset of the code changes of this PR, put's it into a different place that is more to his liking and slaps a "positive review" label on it. He did the same with https://github.com/sagemath/sage/pull/37452, and insisted that it's "positively review" even after I've removed that label (and explained in https://github.com/sagemath/sage/pull/36503#discussion_r1501810781 why the new location is not working for me).

@roed314 @williamstein sorry to bother you with this, but my understanding is that this is a clear misuse of the "positive review" label and, since done multiple times on purpose and not by accident, an abuse of his admin/triage rights. Speaking of abuse of admin rights, Matthias continues to flag my review comments as "resolved" - sometimes without even a response to the question (https://github.com/sagemath/sage/pull/37351#pullrequestreview-1887095019). Normal users without admin rights cannot mark them as "unresolved" again, effectively hiding them from other reviewers. This problematic conduct has been mentioned to Matthias at multiple instances before (and if my memory serves me well even by members of the sage abuse team).

**Comment by dimpase** [comment]
@mkoeppe has blocked me on GH -  while this is sorted out, we can proceed. And if it's not sorted out, we can block him, why not...

**Review by dimpase — APPROVED** [review]

**Comment by roed314** [comment]
The Sage Code of Conduct Committee is setting this back to needs review for now based on it being part of @dimpase's reaction to our decision on #36181, #36561, #36676, #36951, #36964, #37351, #36999, and #36941.  It can, of course, be granted positive review at a future date, and will be subject to the policy on disputed PRs as normal.

**Comment by dimpase** [comment]
is this still a disputed PR?

**Comment by mkoeppe** [comment]
This PR has obviously neither seen sufficient development, nor any meaningful review. It cannot be merged in this form.

It creates a `pyproject.toml` in the SAGE_ROOT, which I have objected to elsewhere. 
- it declares a `build-backend = "setuptools.build_meta"`, which obviously does not work -- there is nothing that backs it, neither a setup.cfg, setup.py, nor [tools.setuptools] information. 
- it duplicates version information, violating our long-standing policy and practice of single-sourcing version information.

To summarize: -1, this is deliberate noise. please close.

**Comment by dimpase** [comment]
we don't really practice "single source version information", it can't therefore "violate" it.

**Comment by mkoeppe** [comment]
> we don't really practice "single source version information", it can't therefore "violate" it.

@dimpase That's a meaningless comment. It seems that you are still confused about our version files even after the documentation improvements in
- https://github.com/sagemath/sage/pull/37401

**Comment by mkoeppe** [comment]
For anyone interested in improving how we source versions: This requires refactoring, as done in:
- https://github.com/sagemath/sage/pull/37430 (needs review)

**Comment by saraedum** [comment]
I see merit in having such a list of dependencies for use cases such as https://github.com/sagemath/sage/pull/37447. I also understand that it's not ideal to have such a file at the top level. Maybe there's a version of this that allows #37447 without putting a `pyproject.toml` into the top-level directory?

**Comment by mkoeppe** [comment]
> I see merit in having such a list of dependencies for use cases such as #37447. I also understand that it's not ideal to have such a file at the top level. Maybe there's a version of this that allows #37447 without putting a `pyproject.toml` into the top-level directory?

@saraedum Yes, I prepared such an unbundled version in #37482.

**Comment by mkoeppe** [comment]
@saraedum @roed314 @jhpalmieri @jplab @VivianePons 
Above (in https://github.com/sagemath/sage/pull/37446#issuecomment-1965521148, https://github.com/sagemath/sage/pull/37446#issuecomment-1965696397) and elsewhere, Tobias and Dima have been attempting to mislead the audience by reframing
- normal Maintainer actions such as creating focused PRs by cherry-picking from unfocused branches (which increases development velocity and is suitable for resolving conflicts), 
- blocking an abusive user on GitHub (a last-resort action)

... as hostile or even abusive actions. 

What is needed are affirmative steps that help repair the damage done by this manipulative reframing.

**Comment by dimpase** [comment]
I strongly object to the framing of proposed alternatives to @mkoeppe's ideas about the directions of Sage project as abuse directed at him.

**Comment by tobiasdiez** [comment]
Setting this to positive review based on:
👍 me (as the author)
👍 @dimpase 
👎 @mkoeppe

**Comment by culler** [comment]
The primary objection to this PR is that is adds a pyproject.toml file in SAGE_ROOT.  The explanation for why this was done is:
> We add this information to a new pyproject.toml file in the root to be independent of the disputed PRs: ...

which seems to not have to do with the design, but rather with the PR review process.

There is some consensus that the ideas in this PR are useful.  There is some evidence that they have not been fully worked out yet.  And attempts to incorporate some of the ideas into a working PR without adding a pyproject.toml file in SAGE_ROOT have been summarily rejected and attacked on what seem to be personal grounds.

I think this is an unfinished proposal which might prove to be useful if people could manage to cooperate.  But there seems to be little hope for that happening.

Because of this state of affairs, I vote -1.  I don't see how merging this PR in the current context could possibly improve Sage.

**Comment by mkoeppe** [comment]
Thanks @culler for the clear analysis. Your vote brings us to:
- in favor: @tobiasdiez (author), @dimpase
- against: @mkoeppe, @culler 

Setting back to "needs review".

**Comment by tobiasdiez** [comment]
> The primary objection to this PR is that is adds a pyproject.toml file in SAGE_ROOT.

So you would prefer if I were to modify the pyproject.toml file in the src directory instead?

**Comment by tobiasdiez** [comment]
@kwankyu As the copyright holder, I'm objecting that my work is used in the form of https://github.com/sagemath/sage/pull/37482 and have hence removed the positive review label there.

If there are things that should be improved, I would like to do this in this PR. For example, I'm still waiting for a response to my earlier question https://github.com/sagemath/sage/pull/37446#issuecomment-2054091431 to one of the reviewers here. Depending on the outcome, this may or may not make #37482 obsolete without creating additional friction for me.

**Comment by saraedum** [comment]
> @kwankyu As the copyright holder, I'm objecting that my work is used in the form of #37482 and have hence removed the positive review label there. Instead, I propose to close #37482 until the discussion in this PR here came to an end.

I'm not a lawyer but I don't think that's how the GPL works. When you published your code here, you made it available under the GPL. You still hold the copyright but that does not give you the right to stop others to use your code within the terms of the GPL.

That said, I think it's obvious that everybody would like to see a consensus on the technical discussion here :)

**Comment by tobiasdiez** [comment]
> You still hold the copyright but that does not give you the right to stop others to use your code within the terms of the GPL.

Sure, but the GPL requires that modifications and the original copyright holder are clearly marked in a prominent place. Perhaps, you would even need to state that sagemath is now a derived work of my fork - which is just absurd. But please not start a legal war on top of the already complicated enough situation.

> That said, I think it's obvious that everybody would like to see a consensus on the technical discussion here :)

I agree. Maybe someone could start by explaining why creating a top-level file pyproject.toml is so controversial. 
As a data point, the following projects have a pyproject file in their root only specifying metadata, ruff or other tools configs - and not being used for the installation of an actual python program. They are all reasonably popular (almost 300.000 stars combined), showing that such a metadata-pyproject file is not uncommon and normally doesn't lead to problems.
- https://github.com/TheAlgorithms/Python/blob/master/pyproject.toml
- https://github.com/nodejs/node/blob/main/pyproject.toml
- https://github.com/HypothesisWorks/hypothesis/blob/master/pyproject.toml
- https://github.com/vladmandic/automatic/blob/master/pyproject.toml

**Comment by roed314** [comment]
> > You still hold the copyright but that does not give you the right to stop others to use your code within the terms of the GPL.
> 
> Sure, but the GPL requires that modifications and the original copyright holder are clearly marked in a prominent place. Perhaps, you would even need to state that sagemath is now a derived work of my fork - which is just absurd. But please not start a legal war on top of the already complicated enough situation.

Here's what I can see about the current status of giving credit:
* Matthias marked you as the author on the [single commit](https://github.com/sagemath/sage/pull/37482/commits) in #37482 
* The text in the PR comment says "Author: @mkoeppe, based on @tobiasdiez's https://github.com/sagemath/sage/pull/37446."
What would you like to see in this regard?

I can't speak to the pyproject.toml question right now, though I hope others will be willing to.

**Comment by mkoeppe** [comment]
> I can't speak to the pyproject.toml question right now, though I hope others will be willing to.

I have explained it in https://github.com/sagemath/sage/pull/37446#issuecomment-1996711954

**Comment by roed314** [comment]
> > I can't speak to the pyproject.toml question right now, though I hope others will be willing to.
> 
> I have explained it in [#37446 (comment)](https://github.com/sagemath/sage/pull/37446#issuecomment-1996711954)

The comment you cite just says "It creates a pyproject.toml in the SAGE_ROOT, which I have objected to elsewhere," but does not include a pointer to where you gave reasons elsewhere.

**Comment by mkoeppe** [comment]
the bullet points below that sentence, no?

**Comment by roed314** [comment]
> the bullet points below that sentence, no?

Ok, I was confused.  Those looked like specific objections about the details of what went into the `pyproject.toml`, rather than objections to having it at all.  For the version information that @tobiasdiez is including in `pyproject.toml` that you don't want to see duplicated, where is it stored now?

**Comment by mkoeppe** [comment]
This should be closed as a duplicate.

**Comment by tobiasdiez** [comment]
The scope of this PR is now largely reduced, only adding a few dependency groups at the end of the pyproject.toml file. I've thus removed the "disputed" flag. (I can also close this PR and open a new one if this is preferred)

**Comment by dimpase** [comment]
I'll be slow to respond in the coming 5-6 days or so, as we're moving across the pond tomorrow.

**Review by dimpase — APPROVED** [review]

**Comment by dimpase** [comment]
lgtm

**Comment by tobiasdiez** [comment]
> lgtm

Thanks!

> ought to be squashed into 1 commit.

Done now.
